File: 354-russian-doll-envelopes/russian-doll-envelopes.cpp
class Solution {
public:
int lengthOfLIS(vector<int>& nums) {
        vector<int> LIS;

        for (int i = 0; i < nums.size(); i++) {
            if (LIS.empty() || LIS.back() < nums[i]) {
                LIS.push_back(nums[i]);
            } else {
                auto it = lower_bound(LIS.begin(), LIS.end(), nums[i]);
                *it = nums[i];
            }
        }

        return LIS.size();
    }
    int maxEnvelopes(vector<vector<int>>& v) {
        
        sort(v.begin(), v.end(), [](const vector<int>& a, const vector<int>& b) {
    if (a[0] == b[0]) return a[1] >b[1];
    return a[0] < b[0];
});

    vector<int>a;
    for(auto it:v)

    {
        a.push_back(it[1]);
    }

return lengthOfLIS(a);

    }
};

File: 354-russian-doll-envelopes/russian-doll-envelopes.py
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        def lis(arr):
            temp=[]
            n=len(arr)
            temp.append(arr[0])
            for i in range(1,n):
                if arr[i]>temp[-1]:
                    temp.append(arr[i])
                else:
                    idx=bisect.bisect_left(temp,arr[i])

                    temp[idx]=arr[i]
            return len(temp)
        
        envelopes.sort(key=lambda x:(x[0],-x[1]))
        h=[]
        for i in range(len(envelopes)):
            h.append(envelopes[i][1])
        return lis(h)
        